src/python/bot/fuzzers/blackbox/engine.py

```diff
@@ -0,0 +1,171 @@
+    Returns:
+      A FuzzOptions object.
+    """
+    app_path = environment.get_value('APP_PATH')
+    app_args = testcase_manager.get_command_line_for_application(
+        get_arguments_only=True).strip()
+    arguments = [
+        f'--app_path={app_path}',
+        f'--app_args={app_args}',
+        f'--input_dir={corpus_dir}',
+    ]
+    return engine.FuzzOptions(corpus_dir, arguments, {})
+
+  def fuzz(self, target_path, options, reproducers_dir, max_time):
+    """Run a fuzzing session.
+    Args:
+      target_path: Path to the target.
+      options: The FuzzOptions object returned by prepare().
+      reproducers_dir: The directory to put reproducers in when crashes
+          are found.
+      max_time: Maximum allowed time for the fuzzing to run.
+   Returns:
+      A FuzzResult object.
+    """
+    os.chmod(target_path, 0o775)
+
+    args = options.arguments
+    args.append(f'--output_dir={reproducers_dir}')
+
+    runner = new_process.UnicodeProcessRunner(target_path)
+    result = runner.run_and_wait(timeout=max_time, additional_args=args)
+
+    crashes = []
+    for testcase_path in os.listdir(reproducers_dir):
+      if not testcase_path.startswith(TESTCASE_PREFIX):
+        continue
+
+      output_path = OUTPUT_PREFIX + testcase_path[len(TESTCASE_PREFIX):]
+      absolute_output_path = os.path.join(reproducers_dir, output_path)
+
+      # If no output was written for a test case, skip it.
+      if not os.path.exists(absolute_output_path):
+        continue
+
+      with open(absolute_output_path, 'r', errors='replace') as handle:
+        output = handle.read()
+
+      # Filter obviously non-crashing test cases. Crashes still follow the
+      # normal flow in fuzz task to ensure that the state should not be ignored
+      # for other reasons, but we don't want to log every test case for the
+      # fuzzers that don't do their own crash processing.
+      state = stack_analyzer.get_crash_data(output)
+      if not state.crash_type:
+        continue
+
+      crash = engine.Crash(
+          os.path.join(reproducers_dir, testcase_path), output, [],
+          int(result.time_executed))
+      crashes.append(crash)
+
+    # TODO(mbarbella): Support stats.
+    stats = {}
+
+    return engine.FuzzResult(result.output, result.command, crashes, stats,
+                             result.time_executed)
+
+  def reproduce(self, target_path, input_path, arguments, max_time):  # pylint: disable=unused-argument
+    """Reproduce a crash given an input.
+    Args:
+      target_path: Path to the target.
+      input_path: Path to the reproducer input.
+      arguments: Additional arguments needed for reproduction.
+      max_time: Maximum allowed time for the reproduction.
+    Returns:
+      A ReproduceResult.
+    """
+    os.chmod(target_path, 0o775)
+
+    app_path = environment.get_value('APP_PATH')
+    app_args = testcase_manager.get_command_line_for_application(
+        get_arguments_only=True).strip()
```

Comment:
Why strip ?

Reply:
Because of how get_command_line_for_application is implemented we're left with extra whitespace. It's not usually a problem since we won't add extra whitespace after the testcase path, but if we don't we might be left with an intermediate string like "-a -b ". It didn't seem like a good idea to pass the trailing whitespace to the fuzzer scripts, but honestly it probably wouldn't hurt to include it.